### PR TITLE
Java Solution for Why Did the Cow Cross the Road (Silver February 2017)

### PR DESCRIPTION
**Pull request by brad182**
If someone can collapse that IO template in the C++ code that would be great I'm not sure how to do that

### CODE REVIEW

**Comment by SansPapyrus683** [comment]
or even better you could just detemplify the code

File: solutions/usaco-714.mdx [hunk 1]
```diff
@@ -238,3 +241,57 @@ int main() {
...
+		for (int x = 0; x < numChickens; x++) {
+			chicken[x] = io.nextInt();
+		}
+		for (int x = 0; x < numCows; x++){
+			a[x] = io.nextInt();
+			b[x] = io.nextInt();
+		}
+		Arrays.sort(chicken);
+		int count = 0;
+
```
Comment (SansPapyrus683) [inline solutions/usaco-714.mdx:272]:
```suggestion
		
		int count = 0;

File: solutions/usaco-714.mdx [hunk 2]
```diff
@@ -238,3 +241,57 @@ int main() {
...
+		for (int x = 0; x < numChickens; x++) { // loop through each chicken
+			int smallest = -1;
+			// find the cow that ends the soonest that the chicken is in its range
+			for(int y = 0; y < numCows; y++){
```
Comment (SansPapyrus683) [inline solutions/usaco-714.mdx:276]: you put spaces between the for and the paren block up there
image
why not do it here as well

File: solutions/usaco-714.mdx [hunk 3]
```diff
@@ -238,3 +241,57 @@ int main() {
...
+public class WhyDidTheCowCrossTheRoad {
+	public static void main (String[] args) throws IOException {
+		Kattio io = new Kattio("helpcross");
+
+		int numChickens = io.nextInt();
+		int numCows = io.nextInt();
+
+		int[] chicken = new int[numChickens]; // where the chicken are located
+		int[] a = new int[numCows];
+		int[] b = new int[numCows];
```
Comment (SansPapyrus683) [inline solutions/usaco-714.mdx:260]: more explicit names would be nice

File: solutions/usaco-714.mdx (hunk 1 above)
Comment (SansPapyrus683) [reply solutions/usaco-714.mdx:272]: also maybe rename this to `crossed`?

File: solutions/usaco-714.mdx [hunk 4]
```diff
@@ -238,3 +241,57 @@ int main() {
...
 }
 ```
+</CPPSection>
+
+<JavaSection>
+```java
+import java.util.*;
+import java.io.*;
+
+public class WhyDidTheCowCrossTheRoad {
```
Comment (SansPapyrus683) [inline solutions/usaco-714.mdx:251]:
```suggestion
public class HelpCross {
```

File: solutions/usaco-714.mdx [hunk 5]
```diff
@@ -238,3 +241,57 @@ int main() {
...
+	public static void main (String[] args) throws IOException {
+		Kattio io = new Kattio("helpcross");
+
+		int numChickens = io.nextInt();
+		int numCows = io.nextInt();
+
+		int[] chicken = new int[numChickens]; // where the chicken are located
+		int[] a = new int[numCows];
+		int[] b = new int[numCows];
+		int[] marked = new int[numCows];
```
Comment (SansPapyrus683) [inline solutions/usaco-714.mdx:261]: if i'm reading the code correctly can't this just be a bool array instead of an int array

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/usaco-714.mdx (hunk 4 above)
Comment (HarryWangATX) [reply solutions/usaco-714.mdx:251]: (optinal)

File: solutions/usaco-714.mdx (hunk 3 above)
Comment (HarryWangATX) [reply solutions/usaco-714.mdx:260]: ^, no 1 letter names.

**Comment by HarryWangATX** [comment]
> or even better you could just detemplify the code

bruh xD

File: solutions/usaco-714.mdx [hunk 6]
```diff
@@ -238,3 +241,57 @@ int main() {
...
+
+		for (int x = 0; x < numChickens; x++) {
+			chicken[x] = io.nextInt();
+		}
+		for (int x = 0; x < numCows; x++){
```
Comment (nathangong) [inline solutions/usaco-714.mdx:266]:
```suggestion
		for (int x = 0; x < numCows; x++) {
```

File: solutions/usaco-714.mdx (hunk 2 above)
Comment (nathangong) [inline solutions/usaco-714.mdx:276]:
```suggestion
			for (int y = 0; y < numCows; y++) {
```

File: solutions/usaco-714.mdx [hunk 7]
```diff
@@ -238,3 +241,57 @@ int main() {
...
+		Arrays.sort(chicken);
+		int count = 0;
+
+		for (int x = 0; x < numChickens; x++) { // loop through each chicken
+			int smallest = -1;
+			// find the cow that ends the soonest that the chicken is in its range
+			for(int y = 0; y < numCows; y++){
+				// the cow has not crossed the road yet and the 
+				// chicken's position is able to help the cow cross
+				if(marked[y] == 0 && chicken[x] >= a[y] && chicken[x] <= b[y]){
```
Comment (nathangong) [inline solutions/usaco-714.mdx:279]:
```suggestion
				if (marked[y] == 0 && chicken[x] >= a[y] && chicken[x] <= b[y]) { 
```

File: solutions/usaco-714.mdx [hunk 8]
```diff
@@ -238,3 +241,57 @@ int main() {
...
+		int count = 0;
+
+		for (int x = 0; x < numChickens; x++) { // loop through each chicken
+			int smallest = -1;
+			// find the cow that ends the soonest that the chicken is in its range
+			for(int y = 0; y < numCows; y++){
+				// the cow has not crossed the road yet and the 
+				// chicken's position is able to help the cow cross
+				if(marked[y] == 0 && chicken[x] >= a[y] && chicken[x] <= b[y]){ 
+					if(smallest < 0 || b[y] < b[smallest]){
```
Comment (nathangong) [inline solutions/usaco-714.mdx:280]:
```suggestion
					if (smallest < 0 || b[y] < b[smallest]) {
```

**Review by nathangong — CHANGES_REQUESTED** [review]
Make sure to maintain consistent spacing throughout your code

**Review by SansPapyrus683 — APPROVED** [review]
i like this code

**Review by nathangong — APPROVED** [review]

**Review by dongliu0426 — CHANGES_REQUESTED** [review]
add

```mdx
# Implementation

**Time Complexity**: $\mathcal O(...)$
```

before 
```mdx
<LanguageSection>
```

File: solutions/usaco-714.mdx [hunk 9]
```diff
@@ -11,6 +11,13 @@ There is more than one possible greedy strategy (though all of them involve
 sorting the animals by time). Here, we follow the approach described by the
 analysis.
 
+## Implementation
+
+**Time Complexity**: $\mathcal O(N*C)$
```
Comment (maggie-j-liu) [inline solutions/usaco-714.mdx:16]:
```suggestion
**Time Complexity:** $\mathcal O(N \cdot C)$
```

File: solutions/usaco-714.mdx [hunk 10]
```diff
@@ -238,3 +245,55 @@ int main() {
...
+		
+		int count = 0;
+		for (int x = 0; x < numChickens; x++) { // loop through each chicken
+			int smallest = -1;
+			// find the cow that ends the soonest that the chicken is in its range
+			for (int y = 0; y < numCows; y++) {
+				// the cow has not crossed the road yet and the 
+				// chicken's position is able to help the cow cross
+				if (! marked[y] && chicken[x] >= earliest[y] && chicken[x] <= latest[y]) { 
+					if(smallest < 0 || latest[y] < latest[smallest]){
```
Comment (maggie-j-liu) [inline solutions/usaco-714.mdx:282]:
```suggestion
					if (smallest < 0 || latest[y] < latest[smallest]) {
```

File: solutions/usaco-714.mdx [hunk 11]
```diff
@@ -238,3 +245,55 @@ int main() {
...
+</CPPSection>
+
+<JavaSection>
+```java
+import java.util.*;
+import java.io.*;
+
+public class HelpCross {
+	public static void main (String[] args) throws IOException {
+		Kattio io = new Kattio("helpcross");
```
Comment (maggie-j-liu) [inline solutions/usaco-714.mdx:257]: add kattio codesnip

**Review by maggie-j-liu — CHANGES_REQUESTED** [review]

File: solutions/usaco-714.mdx (hunk 11 above)
Comment (brad182) [reply solutions/usaco-714.mdx:257]: good catch

**Review by maggie-j-liu — APPROVED** [review]

**Review by dongliu0426 — APPROVED** [review]

**Pull request merged**